Utility trait for stats-based skipping logic

scovich opened the pull request:
Parquet footer stats allow data skipping, very similar to Delta file stats. Except parquet isn't quite as convenient to work with and arrow-parquet doesn't even try to help (it can't, because arrow-compute expressions are opaque, so there's no way to traverse and rewrite them into stats-based skipping predicates).

We implement row group skipping support by traversing the same push-down predicate that delta-kernel already uses to extract a for Delta file skipping predicate. But instead of rewriting the expression, we evaluate it bottom-up (no-copy, O(n) work where n is the number of nodes in the expression). 

This PR does not attempt to actually incorporate the new skipping logic into the default reader. That (plus testing the integration) should likely be a follow-up PR.

OussamaSaoudi-db commented on kernel/src/engine/arrow_footer_skipping.rs:
```diff
@@ -0,0 +1,359 @@
+        };
+        if exprs.iter().any(|v| v.is_some_and(|v| v == dominator)) {
+            Some(dominator)
+        } else if exprs.iter().any(|e| e.is_none()) {
```
I may be missing something, but why is this `any` instead of `all`?

OussamaSaoudi-db replied:
Aha, thank you!

OussamaSaoudi-db commented on kernel/src/engine/arrow_footer_skipping.rs:
```diff
@@ -0,0 +1,359 @@
+}
+
+pub(crate) fn compute_field_indices(
+    fields: &[ColumnDescPtr],
```
I have similar functionality in my [PR](https://github.com/delta-incubator/delta-kernel-rs/pull/349/files), but I just need to extract the columns (I don't have access to a `&[ColumnDescPtr]`). It may be a good idea to split out the `expression_to_column` part so I could reuse your implementation once you merge.

nicklan commented on kernel/src/engine/arrow_footer_skipping.rs:
```diff
@@ -0,0 +1,359 @@
+        match op {
+            Equal => skipping_eq(inverted),
+            NotEqual => skipping_eq(!inverted),
+            LessThan => self.partial_cmp_min_stat(&col, val, Ordering::Less, inverted),
```
if we re-wrote all the `partial_cmp_[min/max]_stat` calls like:
```rust
 LessThan => partial_cmp_scalars(
                &self.get_min_stat_value(&col, &val.data_type())?,
                val,
                Ordering::Less,
                inverted,
            ),
```

We could remove those functions. It's a little more code at the call site though.

scovich replied:
Yeah, I was trying to reduce redundancy as much as possible. There's too much anyway that can't be removed.

nicklan reviewed (commented):
nice, this is really cool. Had a few comments but overall looks great.

hope the unit tests aren't too much of pain to write :)

scovich commented on kernel/src/engine/arrow_footer_skipping.rs:
```diff
@@ -0,0 +1,359 @@
+        .enumerate()
+        .filter_map(|(index, row_group)| {
+            // We can only skip a row group if the filter is false (true/null means keep)
+            let keep = !matches!(RowGroupFilter::apply(filter, row_group), Some(false));
+            keep.then_some(index)
```
I'm not sure those are equivalent? The intent is to keep _unless_ it's `Some(false)`. So `None` and `Some(true)` should both produce the same result. Maybe this?
```suggestion
            let keep = !RowGroupFilter::apply(filter, row_group).is_some_and(|v| !v);
            keep.then_some(index)
```

scovich replied:
This would be a nice place to use `Option::is_none_or`, but that's not stable rust yet:
```suggestion
            RowGroupFilter::apply(filter, row_group).is_none_or(|v| v).then_some(index)
```

hntd187 replied:
> This would be a nice place to use `Option::is_none_or`, but that's not stable rust yet:

You could use just a let chain 
```rust
if let Some(false) = RowGroupFilter::apply(filter, row_group) {
  Some(index)
} else {
  None
}
```

scovich replied:
This code has shifted around quite a bit. I still use `matches!(...)` macro, but the result is a plain bool now instead of `Option<bool>`:
```rust
mpl<'a> RowGroupFilter<'a> {
    /// Applies a filtering expression to a row group. Return value false means to skip it. 
    fn apply(filter: &Expression, row_group: &'a RowGroupMetaData) -> bool {
        let field_indices = compute_field_indices(row_group.schema_descr().columns(), filter);
        let result = Self {
            row_group,
            field_indices,
        }
        .apply_sql_where(filter);
        !matches!(result, Some(false))
    }
```

What do you think?

scovich commented on kernel/src/engine/arrow_footer_skipping.rs:
```diff
@@ -0,0 +1,359 @@
+            (TimestampNtz, _) => None?, // TODO: Int96 timestamps
+            (Decimal(..), _) => None?,  // TODO: Decimal (Int32, Int64, FixedLenByteArray)
+        };
+        Some(value)
+    }
+
+    fn get_max_stat_value(&self, col: &ColumnPath, data_type: &DataType) -> Option<Scalar> {
+        use PrimitiveType::*;
+        let value = match (data_type.as_primitive_opt()?, self.get_stats(col)?) {
+            (String, Statistics::ByteArray(s)) => s.max_opt()?.as_utf8().ok()?.into(),
+            (String, Statistics::FixedLenByteArray(s)) => s.max_opt()?.as_utf8().ok()?.into(),
+            (String, _) => None?,
+            (Long, Statistics::Int64(s)) => s.max_opt()?.into(),
+            (Long, Statistics::Int32(s)) => (*s.max_opt()? as i64).into(),
+            (Long, _) => None?,
+            (Integer, Statistics::Int32(s)) => s.max_opt()?.into(),
+            (Integer, _) => None?,
+            (Short, Statistics::Int32(s)) => (*s.max_opt()? as i16).into(),
+            (Short, _) => None?,
+            (Byte, Statistics::Int32(s)) => (*s.max_opt()? as i8).into(),
+            (Byte, _) => None?,
+            (Float, Statistics::Float(s)) => s.max_opt()?.into(),
+            (Float, _) => None?,
+            (Double, Statistics::Double(s)) => s.max_opt()?.into(),
+            (Double, _) => None?,
+            (Boolean, Statistics::Boolean(s)) => s.max_opt()?.into(),
+            (Boolean, _) => None?,
+            (Binary, Statistics::ByteArray(s)) => s.max_opt()?.data().into(),
+            (Binary, Statistics::FixedLenByteArray(s)) => s.max_opt()?.data().into(),
+            (Binary, _) => None?,
+            (Date, Statistics::Int32(s)) => Scalar::Date(*s.max_opt()?),
+            (Date, _) => None?,
+            (Timestamp, Statistics::Int64(s)) => Scalar::Timestamp(*s.max_opt()?),
+            (Timestamp, _) => None?, // TODO: Int96 timestamps
+            (TimestampNtz, Statistics::Int64(s)) => Scalar::TimestampNtz(*s.max_opt()?),
+            (TimestampNtz, _) => None?, // TODO: Int96 timestamps
+            (Decimal(..), _) => None?,  // TODO: Decimal (Int32, Int64, FixedLenByteArray)
+        };
+        Some(value)
+    }
```
> Was hoping to simplify... but you can't... because `ParquetValueType` isn't publicly exported :(

Exactly... so annoying. I even thought about going `Statistics.[min|max]_bytes_opt`, but it would have been ugly.

As for the suggestion, I thought about doing it that way, but it seemed cleaner (easier to tell what's going on) with the current grouping.

nicklan replied:
Fair enough, it's a few more lines but agree it's a bit easier to follow.

scovich commented on kernel/src/engine/arrow_footer_skipping.rs:
```diff
@@ -0,0 +1,359 @@
+        exprs: &[Expression],
+        inverted: bool,
+    ) -> Option<bool> {
+        let exprs: Vec<_> = exprs
```
Noted, but this code will anyway change a lot -- unit tests uncovered bugs.

nicklan commented on kernel/src/engine/arrow_footer_skipping.rs:
```diff
@@ -0,0 +1,359 @@
+        .filter_map(|(index, row_group)| {
+            // We can only skip a row group if the filter is false (true/null means keep)
+            let keep = !matches!(RowGroupFilter::apply(filter, row_group), Some(false));
+            keep.then_some(index)
```
Right, my bad missing the `None` case (which yours also misses btw @hntd187)

```suggestion
            RowGroupFilter::apply(filter, row_group).unwrap_or(true).then_some(index)
```

hntd187 replied:
Ryan noted that Some(true) and None produced the same results so I assumed it would be okay basically there is only one success case Some(false), unless I misunderstood?

scovich commented:
I just pushed lots of changes:
* Added complete set of tests for the data skipping logic
* Doc comments for everyone
* Significant changes and additions to the data skipping logic itself (partly to fix bugs the tests uncovered)
* Split the code into two files to match the structure that emerged. 

At this point, the PR is no longer WIP. We can decide whether wiring up the row group skipping should be done in this PR or as follow-up work.

scovich commented on kernel/src/engine/parquet_row_group_skipping.rs:
```diff
@@ -0,0 +1,183 @@
+            Expression::Column(name) => {
+                columns.insert(col_name_to_path(name));
+            }
+            Expression::Struct(fields) => {
+                for field in fields {
+                    recurse(field, columns);
+                }
+            }
```
We technically don't need this one, because the skipping logic ignores `Struct` expressions... but I don't know that the restriction is fundamental, so somebody might choose to implement it some day?

scovich commented:
This [build/coverage failure](https://github.com/delta-incubator/delta-kernel-rs/actions/runs/11065218985/job/30744255393?pr=357) doesn't look good?
```
The provided token has expired. Request signature expired at: 2024-09-27T05:46:37+00:00
```

zachschuermann commented:
Ah i'll look into the coverage token issue tomorrow!

scovich commented:
> Ah i'll look into the coverage token issue tomorrow!

Must have been some weird timing race -- a retry succeeded immediately.

hntd187 reviewed (approved):
Those are some impressive tests I have to say. I'm good with all of this, my preference though is that we wire it up in this PR otherwise it's just dead code until that PR lands and we know priorities change and such. But I don't have a strong enough opinion to hold up.

scovich commented:
> my preference though is that we wire it up in this PR otherwise it's just dead code until that PR lands and we know priorities change and such. But I don't have a strong enough opinion to hold up.

I started down that path, but it's a big enough change of its own (and needing tests of its own) that I ended up swinging the other way. All the actual parquet reader logic has been split into a separate PR https://github.com/delta-incubator/delta-kernel-rs/pull/362, and this PR is now completely self-contained and tested. We can keep iterating on the other PR while this one merges.

nicklan commented on kernel/src/engine/parquet_stats_skipping.rs:
```diff
@@ -0,0 +1,1267 @@
+            (Column(col), Literal(val)) => (op, col, val),
+            (Literal(val), Column(col)) => (op.commute()?, col, val),
+            (Literal(a), Literal(b)) => return Self::apply_binary_scalars(op, a, b),
+            _ => None?, // unsupported combination of operands
```
minor point, but do you find `None?` easier to read than `return None` is?

hntd187 replied:
I personally find `None?` to be a bit more confusing to read, maybe I'm reading too much into the natural language of it, but it just seems like you aren't sure if it's none or not. I know that's not what it literally means as it's shorthand, but my point is I do a double take on what it means the first time I see it.

scovich replied:
I can switch it back. Agree it's a bit too "clever"

nicklan commented on kernel/src/engine/parquet_stats_skipping.rs:
```diff
@@ -0,0 +1,1267 @@
+            Some(skip != inverted)
+        };
+        match op {
+            // Given `col == val`:
```
thanks for the comments here, will be useful when debugging :)

nicklan commented on kernel/src/engine/parquet_stats_skipping.rs:
```diff
@@ -0,0 +1,1267 @@
+            // Keep if `NOT(val > max)` implies
+            // Keep if `NOT(max < val)`
+            GreaterThanOrEqual => self.partial_cmp_max_stat(&col, val, Ordering::Less, true),
+            _ => None, // unsupported operation
```
Can we log that we found an unsupported op

scovich replied:
Updated them all.

nicklan commented on kernel/src/engine/parquet_stats_skipping.rs:
```diff
@@ -0,0 +1,1267 @@
+    ColumnPath::new(col.split('.').map(|s| s.to_string()).collect())
+}
+
+#[cfg(test)]
```
given the size of these tests, i think it might be best to make the file tree:
```
kernel/src/engine/
  ...
  parquet_stats_skipping.rs
  parquet_stats_skipping/
    tests.rs
```

And then here you can just do:
```rust
#[cfg(test)]
mod tests;
```

scovich replied:
ah nice... I wanted to do something like that but wasn't sure how.

nicklan commented on kernel/src/engine/parquet_stats_skipping.rs:
```diff
@@ -0,0 +1,1267 @@
+    fn test_binary_lt_ge() {
+        use BinaryOperator::*;
+
+        const LO: Scalar = Scalar::Long(1);
```
Can we use some other scalar types between say the eq_be tests and the lt_gt ones? Probably should be minimal change here but we can exercise the scalar stuff for more than just `Long`

scovich replied:
This test is only trying to verify the conversion regular predicate to data skipping predicate. The different scalar comparisons (by both matched and mismatched types) are already exercised thoroughly by `test_binary_scalars`. 

Is there a particular corner case you worry about, that would make the data skipping code misbehave based on the type of scalar involved?

scovich replied:
Update: Added a negative test, where the literal and column types mismatch.

nicklan reviewed (commented):
yeah, wow that's quite a test suite :)

left a few comments but I'm generally good with merging this as is and then wiring it up later.

Higher level comment. It's great that you've extracted out the stuff needed from the engine so that this PR doesn't need to reference arrow. We have a few other things (and will have more for say `variant`) that fall into this category of "you might want this in your engine no matter what your data format is, so here's some utilities to help you, just implement these traits".

I'm wondering if we want to think about creating a module for that either as a sub-module of `engine` (`engine/utils`?), or as an `engine_utils` standalone module. Probably not in this PR though.

scovich commented on kernel/src/engine/parquet_stats_skipping.rs:
```diff
@@ -0,0 +1,1267 @@
+//! An implementation of data skipping that leverages parquet stats from the file footer.
+use crate::expressions::{BinaryOperator, Expression, Scalar, UnaryOperator, VariadicOperator};
+use crate::schema::DataType;
+use parquet::schema::types::ColumnPath;
```
From https://github.com/delta-incubator/delta-kernel-rs/pull/357#pullrequestreview-2338728716:
> It's great that you've extracted out the stuff needed from the engine so that this PR doesn't need to reference arrow. 

We do still have this one dependency on arrow-parquet [ColumnPath](https://arrow.apache.org/rust/parquet/schema/types/struct.ColumnPath.html). But we already knew we needed to define a similar struct in order to support nested column paths. Once that struct exists, we can use it instead and push the dependency back into the concrete implementation that anyway has to know about arrow-parquet.

scovich replied:
(also added it as a code comment)

zachschuermann replied:
added https://github.com/delta-incubator/delta-kernel-rs/issues/379

scovich commented:
re
> Those are some impressive tests I have to say.
and
> wow that's quite a test suite :)

I know from past bad experience that this is _NOT_ fun stuff to debug. Better to test it near-exhaustively up front and hopefully save some of that pain later.

nicklan reviewed (approved):
lgtm. thanks!

zachschuermann commented on kernel/src/expressions/mod.rs:
```diff
@@ -228,6 +229,10 @@ impl Expression {
         Self::Literal(value.into())
     }
 
+    pub fn null_literal(data_type: DataType) -> Self {
+        Self::Literal(Scalar::Null(data_type))
+    }
```
ah found it - I can use this in the write PR :)
